### PR TITLE
Collect compiler and runner versions before compilation.

### PR DESCRIPTION
**Pull request by meisterT**
Progress towards https://github.com/DOMjudge/domjudge/issues/1241

Each language has canonical information while we store the latest info per judgehost in a separate table.
The information from the judgehost is currently just informative, a next step would be to actually disable judging from this judgehost in case of a version mismatch.

The canonical information can be exposed to the teams via a config option.

### CODE REVIEW

**Comment by meisterT** [comment]
Examples team's page

image

**Comment by meisterT** [comment]
subset of the jury page
image

File: webapp/src/Controller/API/JudgehostController.php [hunk 1]
```diff
@@ -1166,6 +1168,121 @@ public function getFilesAction(
...
+            $ret['compiler_version_command'] = $language->getCompilerVersionCommand();
+        }
+        if (!empty($language->getRunnerVersionCommand())) {
+            $ret['runner_version_command'] = $language->getRunnerVersionCommand();
+        }
+        return $ret;
+    }
+
+    #[IsGranted(new Expression("is_granted('ROLE_JURY') or is_granted('ROLE_JUDGEHOST')"))]
+    #[Rest\Put('/check_versions/{judgetaskid}')]
```
Comment (nickygerritsen) [inline webapp/src/Controller/API/JudgehostController.php:1207]: Would be nice (also for Mart 😛 ) if we document here with OA tags what the body must contain

File: webapp/src/Controller/Team/LanguageController.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#[Route(path: '/team/languages')]
+class LanguageController extends BaseController
+{
+    public function __construct(
+        protected readonly ConfigurationService   $config,
+        protected readonly EntityManagerInterface $em
+    ) {}
+
+    #[Route(path: '/', name: 'team_languages')]
+    public function langAction(): Response
```
Comment (nickygerritsen) [inline webapp/src/Controller/Team/LanguageController.php:28]: I'd prefer `languagesAction`

File: webapp/src/DataFixtures/DefaultData/LanguageFixture.php [hunk 3]
```diff
@@ -25,28 +25,28 @@ public function load(ObjectManager $manager): void
...
+            ['csharp', 'csharp',     'C#',          ['csharp', 'cs'],            false, null,         false,  true,   1,     'csharp',            '', ''],
+            ['f95',    'f95',        'Fortran',     ['f95', 'f90'],              false, null,         false,  true,   1,     'f95',               '', ''],
+            ['hs',     'haskell',    'Haskell',     ['hs', 'lhs'],               false, null,         false,  true,   1,     'hs',                '', ''],
+            ['java',   'java',       'Java',        ['java'],                    false, 'Main class', true,   true,   1,     'java_javac_detect', 'javac -version', 'java -version'],
+            ['js',     'javascript', 'JavaScript',  ['js'],                      false, 'Main file',  false,  true,   1,     'js',                '', ''],
+            ['lua',    'lua',        'Lua',         ['lua'],                     false, null,         false,  true,   1,     'lua',               '', ''],
+            ['kt',     'kotlin',     'Kotlin',      ['kt'],                      true,  'Main class', false,  true,   1,     'kt',                '', ''],
```
Comment (nickygerritsen) [inline webapp/src/DataFixtures/DefaultData/LanguageFixture.php:39]: Should we at least add kotlin version commands?

File: webapp/src/DataFixtures/DefaultData/LanguageFixture.php [hunk 4]
```diff
@@ -25,28 +25,28 @@ public function load(ObjectManager $manager): void
...
+            ['pas',    'pascal',     'Pascal',      ['pas', 'p'],                false, 'Main file',  false,  true,   1,     'pas',               '', ''],
+            ['pl',     'pl',         'Perl',        ['pl'],                      false, 'Main file',  false,  true,   1,     'pl',                '', ''],
+            ['plg',    'prolog',     'Prolog',      ['plg'],                     false, 'Main file',  false,  true,   1,     'plg',               '', ''],
+            ['py3',    'python3',    'Python 3',    ['py'],                      false, 'Main file',  true,   true,   1,     'py3',               'python3 --version', 'pypy3 --version'],
```
Comment (nickygerritsen) [inline webapp/src/DataFixtures/DefaultData/LanguageFixture.php:43]: This is a TODO, to not use pypy3 AND python3 I guess

File: webapp/src/Entity/ImmutableExecutable.php [hunk 5]
```diff
@@ -41,6 +41,18 @@ class ImmutableExecutable
     #[ORM\Column(length: 32, nullable: true, options: ['comment' => 'hash of the files'])]
     private ?string $hash;
 
+    #[ORM\Column(type: 'blobtext', nullable: true, options: ['comment' => 'Compiler version'])]
```
Comment (nickygerritsen) [inline webapp/src/Entity/ImmutableExecutable.php:44]: This can be dropped right?

File: webapp/src/Entity/ImmutableExecutable.php [hunk 6]
```diff
@@ -97,4 +109,48 @@ public function disallowDelete(): never
     {
         throw new RuntimeException('An immutable executable cannot be deleted');
     }
+
```
Comment (nickygerritsen) [inline webapp/src/Entity/ImmutableExecutable.php:112]: These as well.

File: webapp/src/Entity/Language.php [hunk 7]
```diff
@@ -114,6 +114,99 @@ class Language extends BaseApiEntity
...
+
+    #[ORM\Column(type: 'string', length: 255, nullable: true, options: ['comment' => 'Compiler version command'])]
+    private ?string $compilerVersionCommand = null;
+
+    public function getVersions()
+    {
+        return $this->versions;
+    }
+
+    /**
```
Comment (nickygerritsen) [inline webapp/src/Entity/Language.php:138]: Let's drop the useless docblocks.

File: webapp/src/Entity/Version.php [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    )]
+    #[Serializer\Exclude]
+    private string|float|null $lastChangedTime = null;
+
+    public function getVersionid(): ?int
+    {
+        return $this->versionid;
+    }
+
+    /**
```
Comment (nickygerritsen) [inline webapp/src/Entity/Version.php:59]: Let's also drop docblocks here.

File: webapp/templates/jury/executable.html.twig [hunk 9]
```diff
@@ -72,6 +72,39 @@
                         {% endif %}
                     </td>
                 </tr>
+                {% if executable.type == 'compile' %}
```
Comment (nickygerritsen) [inline webapp/templates/jury/executable.html.twig:75]: This is also old.

File: webapp/src/Controller/API/JudgehostController.php [hunk 10]
```diff
@@ -1166,6 +1168,159 @@ public function getFilesAction(
...
+    #[OA\Response(
+        response: 200,
+        description: 'Updates internal versions, does not check them yet.',
+    )]
+    #[OA\Parameter(
+        name: 'hostname',
+        description: 'Checking versions for the given hostname.',
+        in: 'query',
```
[the remark this answers is not in the document]
Comment (meisterT) [reply webapp/src/Controller/API/JudgehostController.php:1230]: Are they? Getting them from the request as other query params?

[the remark this answers is not in the document]
Comment (nickygerritsen) [reply webapp/src/Controller/API/JudgehostController.php:1227]: You are PUTting so it is a body, not in the query. So I think this all should be replaced by:
```php
#[OA\RequestBody(
        required: true,
        content: [
            new OA\MediaType(
                mediaType: 'multipart/form-data',
                schema: new OA\Schema(
                    required: ['hostname', 'compiler', 'runner'],
                    properties: [
                        new OA\Property(
                            property: 'problem',
                            description: 'Checking versions for the given hostname.',
                            type: 'string'
                        ),
                        new OA\Property(
                            property: 'compiler',
                            description: 'Reported compiler version.',
                            type: 'string'
                        ),
                        new OA\Property(
                            property: 'runner',
                            description: 'Reported runner version.',
                            type: 'array',
                            items: new OA\Items(type: 'string', format: 'binary')
                        ),
                    ]
                )
            ),
        ]
    )]
```

[the remark this answers is not in the document]
Comment (meisterT) [reply webapp/src/Controller/API/JudgehostController.php]: thanks

**Review by nickygerritsen — APPROVED** [review]

File: webapp/src/Controller/API/JudgehostController.php [hunk 11]
```diff
@@ -1166,6 +1168,166 @@ public function getFilesAction(
...
+            type: 'object'
+        )
+    )]
+    public function getVersionCommands(string $judgetaskid): array
+    {
+        /** @var JudgeTask $judgeTask */
+        $judgeTask = $this->em->getRepository(JudgeTask::class)
+            ->findOneBy(['judgetaskid' => $judgetaskid]);
+        if (!$judgeTask) {
+            throw new BadRequestHttpException('Unknown judge task with id ' . $judgetaskid);
```
[the remark this answers is not in the document]
Comment (meisterT) [reply webapp/src/Controller/API/JudgehostController.php:1194]: Consistent with the other cases how we handle the missing data in the judgehostcontroller.

**Review by eldering — APPROVED** [review]

**Pull request merged**